org.osgi.test.assertj/src/main/java/org/osgi/test/assertj/promise/PromiseSoftAssertions.java

```diff
@@ -22,17 +22,4 @@
 /**
  * Soft assertions for {@link Promise}s.
  */
-public class PromiseSoftAssertions extends SoftAssertions {
-	/**
-	 * Create a soft assertion for a {@link Promise}.
-	 *
-	 * @param actual The {@link Promise}.
-	 * @param <T> The type of the value contained in the {@link Promise}.
-	 * @return The created soft assertion.
-	 */
-	public <T> PromiseAssert<T> assertThat(Promise<? extends T> actual) {
-		@SuppressWarnings("unchecked")
-		PromiseAssert<T> softly = proxy(PromiseAssert.class, Promise.class, actual);
-		return softly;
-	}
-}
+public class PromiseSoftAssertions extends SoftAssertions implements PromiseSoftAssertionsProvider {}
```

Comment:
Should we not get rid of this class and just have the interface?

Reply:
We certainly can. I kept it for backward compatibility in case anyone was using it (including the test cases). But I suppose we're in pre-release and anyone using it should be ready to handle such incompatible changes.

Reply:
I also think an uber-interface like `OSGiSoftAssertionsProvider` would be a be a good move as we get more of these.

Reply:
I am not super excited by uber types. They are non-cohesive. I think it is better that each package be focused and we don't end up with a dumping ground like java.util.

Reply:
We should keep the `PromiseSoftAssertions` class. I just used it :-)

Reply:
> I am not super excited by uber types. They are non-cohesive. I think it is better that each package be focused and we don't end up with a dumping ground like java.util.

But the uber-types for the assertion entry points are one of the main features of AssertJ! They allow you to type `softly.assertThat().` without having to make sure you've instantiated just the exact correct `SoftAssertionsProvider` for the object that you're asserting on - it's already there. And when you're writing real-world tests (and especially using soft assertions), it is usual that you will want to assert on objects of multiple different types, and that you will want these assertion failures (if any) to be reported in a group in the order that they are executed.

AssertJ itself does not follow the pattern of defining one `SoftAssertionsProvider` per assertion type.

Reply:
> AssertJ itself does not follow the pattern of defining one `SoftAssertionsProvider` per assertion type.

But AssertJ is the the base and a complete entity. Like Java itself. We are just an appendage. From a package versioning point of view, such non-cohesive types create lots of changes to the uber package as each piece changes.

With the provider, it is now super easy for any user of our custom assertions and someone else's custom assertions to easy make what ever uber type makes sense for them, no?

Reply:
> > AssertJ itself does not follow the pattern of defining one `SoftAssertionsProvider` per assertion type.
> 
> But AssertJ is the the base and a complete entity. Like Java itself. We are just an appendage.

We are an appendage, but a complete one. It is natural for OSGi assertion types to be bundled together - literally! (we're not creating an individual bundle for each assertion type).

> From a package versioning point of view, such non-cohesive types create lots of changes to the uber package as each piece changes.
>

I don't see how an uber-class would be a source of too many versioning changes - it only would be a MINOR change when you add a new assertion type (which is infrequent), or a MAJOR change when you remove one (even less frequent).

With the one-per-assertion model, you end up with a lot more soft assertion provider classes that are a potential source of versioning changes to the packages that they live in.

> With the provider, it is now super easy for any user of our custom assertions and someone else's custom assertions to easy make what ever uber type makes sense for them, no?

Well, I suppose it's not too bad - but it's even easier to add one for all of osgi-test than it is to try and add VersionSoftAssertionProvider, BundleSoftAssertionProvider, PromiseSoftAssertionProvider, etc individually. It's easy now because we've only got a few of them, but what about when we've got dozens? Are we going to require every serious user of osgi-test to manually build their own uber-soft assertions entry point in this way?

Reply:
> Are we going to require every serious user of osgi-test to manually build their own uber-soft assertions entry point in this way?

OK, lets try to make our uber type. Where does it live? In the base `org.osgi.test.assertj` package? Or in a new subpackage called ??